truffleruby/src/main/java/org/truffleruby/core/binding/BindingNodes.java

```diff
@@ -212,73 +216,64 @@ public RubyNode coerceToString(RubyNode name) {
             return NameToJavaStringNodeGen.create(name);
         }
 
-
         @Specialization(guards = {
                 "name == cachedName",
-                "!isLastLine(cachedName)",
-                "compatibleFrames(binding, cachedBinding)",
-                "cachedFrameSlot != null"
-        },
-                limit = "getCacheLimit()")
+                "!hiddenVariable(name)",
+                "cachedFrameSlot != null",
+                "getFrameDescriptor(binding) == descriptor"
+        }, limit = "getCacheLimit()")
         public Object localVariableGetCached(DynamicObject binding, String name,
-                                             @Cached("name") String cachedName,
-                                             @Cached("binding") DynamicObject cachedBinding,
-                                             @Cached("findFrameSlotOrNull(binding, name)") FrameSlotAndDepth cachedFrameSlot,
-                                             @Cached("createReadNode(cachedFrameSlot)") ReadFrameSlotNode readLocalVariableNode) {
+                @Cached("name") String cachedName,
+                @Cached("getFrameDescriptor(binding)") FrameDescriptor descriptor,
+                @Cached("findFrameSlotOrNull(binding, name)") FrameSlotAndDepth cachedFrameSlot,
+                @Cached("createReadNode(cachedFrameSlot)") ReadFrameSlotNode readLocalVariableNode) {
             final MaterializedFrame frame = RubyArguments.getDeclarationFrame(Layouts.BINDING.getFrame(binding), cachedFrameSlot.depth);
             return readLocalVariableNode.executeRead(frame);
         }
 
+        @Specialization(guards = {
+                "name == cachedName",
+                "!hiddenVariable(name)",
+                "cachedFrameSlot != null",
+                "getExtrasDescriptorOrNull(binding) == descriptor"
+        }, limit = "getCacheLimit()")
+        public Object extraVariableGetCached(DynamicObject binding, String name,
+                @Cached("name") String cachedName,
+                @Cached("getExtrasDescriptorOrNull(binding)") FrameDescriptor descriptor,
+                @Cached("findExtrasSlotOrNull(binding, name)") FrameSlotAndDepth cachedFrameSlot,
+                @Cached("createReadNode(cachedFrameSlot)") ReadFrameSlotNode readLocalVariableNode) {
+            final MaterializedFrame frame = RubyArguments.getDeclarationFrame(Layouts.BINDING.getExtras(binding),
+                    cachedFrameSlot.depth);
```

Comment:
Good spot on the declaration frame issue, it would have produced an error if asserts were enabled. I've tweaked this now to only get declaration frames if the depth is greater than 0 (occurs with blocks).